### ISSUE TITLE
Email validation regex too strict

### ISSUE THREAD

**Issue by veqryn** [asaskevich/govalidator#131]
The long and complicated regex you are presently using to validate email addresses is way too strict according to the RFC for email addresses.

Below is a list of valid email address, of which your regex matches less than half.
```
blah@gmail.com
test@d.verylongtoplevel
holy..moly.@gmail.com
email+tag@gmail.com
" "@example.com
“Abc\@def”@example.com
“Fred Bloggs”@example.com
“Joe\\Blow”@example.com
“Abc@def”@example.com
customer/department=shipping@example.com
$A12345@example.com
!def!xyz%abc@example.com
_somename@example.com
much.“more\ unusual”@example.com
very.unusual.“@”.unusual.com@example.com
very.“(),:;<>[]”.VERY.“very@\\ "very”.unusual@strange.example.com
!#$%&'*+-/=?^_`{}|~@example.com
Miles.O'Brian@example.com
postmaster@☁→❄→☃→☀→☺→☂→☹→✝.ws
allen@[127.0.0.1]
allen@[IPv6:0:0:1]
root@localhost
john@com
```

I'd recommend changing the regex to just this: `^.+@.+$`

**Comment by annismckenzie** [comment]
I'm against changing the regex to this oversimplistic one – you may add your own struct validation tag or your own custom validation function.

```go
  rxEmailLax = regexp.MustCompile("^.+@.+$")
  govalidator.TagMap["email_lax"] = govalidator.Validator(func(str string) bool {
    return rxEmailLax.MatchString(str)
  })
```

**Comment by veqryn** [comment]
Fair enough.
What about improving the current regex to cover more of the above examples?

**Comment by annismckenzie** [comment]
Let me ask you this: what real-world email address that you currently need to handle failed the validator? The list above is, to be frank, a bit weird.

Looking at them, here's the list with checkboxes – unchecked means it fails validation:
- [x] blah@gmail.com
- [x] test@d.verylongtoplevel
- [ ] holy..moly.@gmail.com
- [x] email+tag@gmail.com
- [ ] " "@example.com
- [ ] “Abc\@def”@example.com
- [ ] “Fred Bloggs”@example.com
- [ ] “Joe\\Blow”@example.com
- [ ] “Abc@def”@example.com
- [x] customer/department=shipping@example.com
- [x] $A12345@example.com
- [x] !def!xyz%abc@example.com
- [x] _somename@example.com
- [ ] much.“more\ unusual”@example.com
- [ ] very.unusual.“@”.unusual.com@example.com
- [ ] very.“(),:;<>[]”.VERY.“very@\\ "very”.unusual@strange.example.com
- [x] !#$%&'*+-/=?^_`{}|~@example.com
- [x] Miles.O'Brian@example.com
- [x] postmaster@☁→❄→☃→☀→☺→☂→☹→✝.ws
- [ ] allen@[127.0.0.1]
- [ ] allen@[IPv6:0:0:1]
- [ ] root@localhost
- [ ] john@com

Which one of those unchecked would you like fixed first? Looking at the regex I see that it's a monster. Will try to check for an alternative as well.

**Comment by annismckenzie** [comment]
Interesting. GitHub tries to autolink these addresses and the validator succeeds in a few where GitHub does not link them as one email address (just parts of it). All of the unchecked ones are not recognized by GitHub either.

**Comment by veqryn** [comment]
I guess this sort of depends on how much more complicated you want the regex to get...
You could probably handle the holy..moly.@gmail.com address without much difficulty, but everything else looks quite tough to handle.
And whether you want to handle the last 4 (the IP based and the domains lacking a dot), might be debatable for some of your users.

**Comment by annismckenzie** [comment]
You're right. To be honest, I have never seen or met someone with an email address looking like the ones above that are unchecked. Emojis, backslashes, … – who remembers this stuff? :D

Oh, one jumped out at me from the corporate world: “Fred Bloggs”@example.com. These I have seen once.

Waiting on @asaskevich to chime in.

**Comment by asaskevich** [comment]
Better to follow RFC 😄 But I'm not sure that this regex will validate valid (as it said in RFC) and invalid (according to RFC) emails.

**Comment by epelc** [comment]
@asaskevich any reason why [`mail.ParseAddress`](https://golang.org/pkg/net/mail/#ParseAddress) isn't being used?

**Comment by dolmen** [comment]
Instead it would be better to use the regexp provided in the HTML5 specification for e-mail input fields:
https://www.w3.org/TR/html5/forms.html#valid-e-mail-address
The HTML5 spec explains why the specification RFC5322 is broken and should not be followed.

**Comment by dolmen** [comment]
@eplc Is `"Alice <alice@example.com>"` really something you want to accept as an e-mail address? Because that's what is allowed by RFC5322 implemented in mail.ParseAddress. https://play.golang.org/p/JrjTc96Yl6

**Comment by epelc** [comment]
@dolmen I was thinking you could get the `Address` field. It's definitely not ideal though.

I wasn't aware html5 had a different spec for email formats.

**Comment by macrael** [comment]
.mil addresses are not recognized. I too suggest a simpler regex for this. I like the advice on this site: http://emailregex.com though the HTML definition mentioned above also looks good.

**Comment by 0sc** [comment]
I see the email validation was updated sometime after this thread. Was there a standard settled for? Currently, it marks entries such as `"]"@bar.com`,`")"@bar.com`, `&@bar.com`, `"@bar.com`, `'@bar.com.`

**Comment by ptman** [comment]
Maybe use the regex used by HTML5 browsers to validate type=email inputs: https://www.w3.org/TR/html5/forms.html#valid-e-mail-address

**Comment by ptman** [comment]
We had a problem with email addresses that had a three component domain, e.g. foo@bar.baz.com. Solved by switching to HTML5 regex.

**Comment by dhax** [comment]
any movement on this default regex? While using a fake data generator during automated testing I came across these valid looking addresses, but all fail the default validation:

rosariokoch@kshlerin.info, eunicehaag@heidenreich.io, paolomonahan@hodkiewicz.name, erickernser@hilll.name,  coythompson@oberbrunner.info, mikaylawaelchi@greenholt.net

**Comment by sergeyglazyrindev** [comment]
Hello guys!
I forked this package cause owner disappeared. Hope, he will be back, but it would be easier to merge these changes back if he is back
Link to my repo: create issue there and we'll discuss it.

**Comment by morigs** [comment]
Also, I get `does not validate as email` error when use gmail-style emails with plus sign (e.g. `some-name+some-tag@gmail.com`)
